doc_internal/fish-riir-plan.md

```diff
@@ -0,0 +1,77 @@
+These is a proposed port of fish-shell from C++ to Rust, and from CMake to cargo or related. This document is high level - see the Development Guide for more details.
+
+## Why Port
+
+- Gain access to more contributors and enable easier contributions. C++ is becoming a legacy language.
+- Free us from the annoyances of C++/CMake, and old toolchains.
+- Ensure fish continues to be perceived as modern and relevant.
+- Unlock concurrent mode (see below).
+
+## Why Rust
+
+- Rust is a systems programming language with broad platform support, a large community, and a relatively high probability of still being relevant in a decade.
+- Rust has a unique strength in its thread safety features, which is the missing piece to enable concurrent mode - see below.
+- Other languages considered:
+  - Java, Python and the scripting family are ruled out for startup latency and memory usage reasons.
+  - Go would be an awkward fit. fork is [quite the problem](https://stackoverflow.com/questions/28370646/how-do-i-fork-a-go-process/28371586#28371586) in Go.
+  - Other system languages (D, Nim, Zig...) are too niche: fewer contributors, higher risk of the language becoming irrelevant.
```

Comment:
(FWIW Java startup time can be fixed by compiling a [native image](https://www.graalvm.org/22.0/reference-manual/native-image/))

Reply:
I don't really feel qualified to comment on this proposal, being new to fish and the languages involved, but as long as the topic is open I would encourage the team to take a second look at D.

- a mature language (22+ years)
- has multiple compilers available (a reference compiler, D is part of GCC, and there is an LLVM frontend)
- is intended to be familiar to C++ users (many C++11 features seem to be inspired by D)
- has excellent interoperability with C++ code

I have no experience with Rust so I can't comment on that. I also am unfamiliar with the concurrent execution problem, but note that D has mature concurrency features.

Reply:
Let me put a swift lid on that (and also swift and any other discussions of other programming languages).

Rust has the momentum, some of us already know it and others are interested in it.

This exists, it is reasonably far along, it works. The decision has essentially been made.

Reply:
@krobelus GraalVM Native Image is almost definitely not a reasonable option here, it *will* not shoot out small binaries. An *empty* class will have a 20 MB binary — A larger application ranging in the hundreds of megabytes (I've personally seen 370+!)